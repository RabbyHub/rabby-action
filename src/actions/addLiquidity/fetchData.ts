import PQueue from 'p-queue';
import {
  AddLiquidityRequireData,
  FetchActionRequiredData,
  ParsedActionData,
} from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import BigNumber from 'bignumber.js';

export const fetchDataAddLiquidity: FetchActionRequiredData<{
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: any;
  token1: any;
}> = async (options, likeAction) => {
  const { actionData, apiProvider, sender, walletProvider, chainId } = options;

  const action =
    likeAction || (<ParsedActionData<'transaction'>>actionData).addLiquidity;

  if (!action || options.type === 'text') {
    return {};
  }

  const { receiver, exchange_rate, token0, token1 } = action;

  const fair_exchange_rate = new BigNumber(token0.price).div(token1.price);

  const diff = +new BigNumber(exchange_rate)
    .minus(fair_exchange_rate)
    .abs()
    .div(fair_exchange_rate)
    .times(100)
    .toFixed(4);

  const receiverInWallet = await walletProvider.hasAddress(receiver);

  const id =
    options.type === 'transaction'
      ? options.tx.to
      : options.actionData.contractId;
  if (!id) {
    return {};
  }

  const result: AddLiquidityRequireData = {
    id,
    protocol: null,
    bornAt: 0,
    rank: null,
    sender,
    receiverInWallet,
    hasInteraction: false,
    receiver,
    poolRate: exchange_rate,
    token0,
    token1,
    marketRate: fair_exchange_rate.toFixed(),
    diff,
  };
  const queue = new PQueue();

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(id, chainId);
    if (!contractInfo) {
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.bornAt = contractInfo.create_at;

      result.protocol = contractInfo.protocol;
    }
  });

  queue.add(async () => {
    const hasInteraction = await apiProvider.hasInteraction(
      sender,
      chainId,
      id
    );
    result.hasInteraction = hasInteraction.has_interaction;
  });

  await waitQueueFinished(queue);

  return result;
};
