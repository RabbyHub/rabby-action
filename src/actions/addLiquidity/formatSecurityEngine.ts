import {
  AddLiquidityRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineAddLiquidity: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId } = options;

    if (!actionData.addLiquidity || !chainId) {
      return {} as any;
    }

    const data = requireData as AddLiquidityRequireData;

    return {
      addLiquidity: {
        receiverMatch:
          data.receiver?.toLowerCase() === data.sender.toLowerCase(),
        diff: data.diff,
      },
    };
  };
