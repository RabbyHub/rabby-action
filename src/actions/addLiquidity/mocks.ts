export const preExecData = {
  abi: {
    func: 'mint',
    params: [
      [
        '0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22',
        '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913',
        10,
        -4970,
        -4960,
        2955759909,
        885199302,
        2940981110,
        880773306,
        '0xE2866F2291B5c5bafD2858B9247f93745108A523',
        1754366820,
        0,
      ],
    ],
  },
  abi_str:
    '{\n    "func": "mint",\n    "params": [\n        [\n            "0x18Bc5bcC660cf2B9cE3cd51a404aFe1a0cBD3C22",\n            "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",\n            10,\n            -4970,\n            -4960,\n            2955759909,\n            885199302,\n            2940981110,\n            880773306,\n            "0xE2866F2291B5c5bafD2858B9247f93745108A523",\n            1754366820,\n            0\n        ]\n    ]\n}',
  balance_change: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 29320,
    gas_limit: 0,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'base',
    chain: 'base',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 3629.55,
    price_24h_change: 0.018385589869839923,
    credit_score: 46503412.60907772,
    is_verified: true,
    is_scam: false,
    is_suspicious: null,
    is_core: true,
    is_wallet: true,
    time_at: null,
    low_credit_score: false,
    amount: 1e-5,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: false,
    error: {
      code: 1002,
      msg: 'execution reverted: execution reverted',
    },
  },
  type_call: {
    action: 'mint',
    contract: '0x827922686190790b37229fd06084350e74485b72',
    contract_protocol_name: 'Aerodrome V3',
    contract_protocol_logo_url:
      'https://static.debank.com/image/project/logo_url/base_aerodrome/f02d753bc321dc8ba480f0424a686482.png',
  },
  trace_id: '663ae6c6c0ac4b1aa5651fadba316de2',
} as any;

export const parseTxData = {
  action: { type: 'cancel_tx' },
  contract_call: null,
  log_id: 441860918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  to: '0x4686fd7c0bc5d06346d416808916ba1dbd3f2336',
  value: '0x0',
  data: '0x',
  gas: '0x0',
  maxFeePerGas: '0xb71b00',
  maxPriorityFeePerGas: '0xb71b00',
  nonce: '0x4',
};
