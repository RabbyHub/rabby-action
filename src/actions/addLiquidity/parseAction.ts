import { ParseAction } from '../../types';
import { TokenItem } from '@rabby-wallet/rabby-api/dist/types';

interface AddLiquidityActionData {
  receiver: string;
  exchange_rate: string;
  min_exchange_rate: string;
  max_exchange_rate: string;
  token0: TokenItem;
  token1: TokenItem;
}

export const parseActionAddLiquidity: ParseAction = (options) => {
  const { data } = options;

  if (data?.type !== 'add_liquidity') {
    return {};
  }

  const {
    receiver,
    exchange_rate,
    min_exchange_rate,
    max_exchange_rate,
    token0,
    token1,
  } = data.data as unknown as AddLiquidityActionData;

  return {
    addLiquidity: {
      receiver,
      exchange_rate,
      min_exchange_rate,
      max_exchange_rate,
      token0,
      token1,
    },
  };
};
