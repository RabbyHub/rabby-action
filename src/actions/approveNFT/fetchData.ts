import PQueue from 'p-queue';
import { ApproveNFTRequireData, FetchActionRequiredData } from '../../types';
import { waitQueueFinished } from '../../utils/waitQueueFinished';
import { fetchHasInteraction } from '../../fetches/fetchHasInteraction';

export const fetchDataApproveNFT: FetchActionRequiredData<{
  spender: string;
}> = async (options, likeAction) => {
  const queue = new PQueue();
  const { sender, apiProvider, chainId, actionData, extraActionDataState } =
    options;
  const action = likeAction || actionData.approveNFT;

  if (!action) {
    return {};
  }

  const result: ApproveNFTRequireData = {
    isEOA: false,
    contract: null,
    riskExposure: 0,
    rank: null,
    bornAt: 0,
    protocol: null,
    isDanger: false,
    tokenBalance: '0',
    hasInteraction: null,
    extraState: {
      hasInteraction: () => undefined,
    },
  };

  queue.add(async () => {
    const contractInfo = await apiProvider.getContractInfo(
      action.spender,
      chainId
    );
    if (!contractInfo) {
      result.isEOA = true;
      result.rank = null;
    } else {
      result.rank = contractInfo.credit.rank_at;
      result.riskExposure = contractInfo.top_nft_spend_usd_value;
      result.bornAt = contractInfo.create_at;
      result.isDanger =
        contractInfo.is_danger.auto || contractInfo.is_danger.edit;
      result.protocol = contractInfo.protocol;
    }

    if (result.isEOA) {
      queue.add(async () => {
        const { desc } = await apiProvider.addrDesc(action.spender);
        result.bornAt = desc.born_at;
      });
    }
  });

  fetchHasInteraction({
    apiProvider,
    sender,
    chainId,
    spender: action.spender,
    extraActionDataState,
    queue,
    result,
  });

  await waitQueueFinished(queue);
  return result;
};
