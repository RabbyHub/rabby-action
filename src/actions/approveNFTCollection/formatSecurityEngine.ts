import {
  ApproveNFTRequireData,
  FormatSecurityEngineContext,
} from '../../types';

export const formatSecurityEngineApproveNFTCollection: FormatSecurityEngineContext =
  async (options) => {
    const { actionData, requireData, chainId, provider } = options;

    if (!actionData.approveNFTCollection || !chainId) {
      return {};
    }
    const data = requireData as ApproveNFTRequireData;
    const { spender } = actionData.approveNFTCollection;
    return {
      collectionApprove: {
        chainId,
        spender,
        isEOA: data.isEOA,
        riskExposure: data.riskExposure,
        deployDays: provider.getTimeSpan(
          Math.floor(Date.now() / 1000) - data.bornAt
        ).d,
        isDanger: !!data.isDanger,
      },
    };
  };
