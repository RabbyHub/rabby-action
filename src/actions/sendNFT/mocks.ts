export const preExecData = {
  abi: {
    func: 'safeTransferFrom',
    params: [
      '0x5853eD4f26A3fceA565b3FBC698bb19cdF6DEB85',
      '0x52DeC422AA4bAE0fB6F2b41db023dFfA9f4A1353',
      440299,
    ],
  },
  abi_str:
    '{\n    "func": "safeTransferFrom",\n    "params": [\n        "0x5853eD4f26A3fceA565b3FBC698bb19cdF6DEB85",\n        "0x52DeC422AA4bAE0fB6F2b41db023dFfA9f4A1353",\n        440299\n    ]\n}',
  balance_change: {
    success: true,
    error: null,
    send_token_list: [],
    receive_token_list: [],
    send_nft_list: [
      {
        id: '7b31129c599fee1f53b8f439c0391072',
        contract_id: '0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
        inner_id: '440299',
        chain: 'arb',
        symbol: 'CUBE',
        name: 'Mission 200K: Arbitrum',
        description: '',
        content_type: 'image_url',
        content:
          'https://static.debank.com/image/arb_nft/local_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
        thumbnail_url:
          'https://static.debank.com/image/arb_nft/thumbnail_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
        total_supply: 1,
        attributes: [
          {
            trait_type: 'Type',
            value: 'Quest',
          },
          {
            trait_type: 'Title',
            value: 'Mission 200K: Arbitrum',
          },
          {
            trait_type: 'Community',
            value: 'Layer3',
          },
          {
            trait_type: 'Tag',
            value: 'DeFi',
          },
          {
            trait_type: 'Difficulty',
            value: 'BEGINNER',
          },
          {
            trait_type: 'Wallet Provider',
            value: 'Rabby Wallet',
          },
        ],
        detail_url:
          'https://opensea.io/assets/0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7/440299',
        collection_id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
        is_erc1155: false,
        is_erc721: true,
        pay_token: null,
        collection: {
          id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
          chain: 'arb',
          name: 'Layer3 CUBE',
          description: null,
          logo_url:
            'https://static.debank.com/image/arb_nft/thumbnail_url/24b44272ad6bf55323dca1add5f520f6/7b78944a6e25d3d2eded3eb6425802a1.png',
          is_verified: true,
          credit_score: 1643474.735188378,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          floor_price: 0.0,
        },
        amount: 1,
      },
    ],
    receive_nft_list: [],
    usd_value_change: 0,
  },
  gas: {
    success: true,
    error: null,
    gas_used: 69241,
    gas_limit: 75585,
    gas_ratio: 1.3,
  },
  is_gnosis: false,
  native_token: {
    id: 'arb',
    chain: 'arb',
    name: 'ETH',
    symbol: 'ETH',
    display_symbol: null,
    optimized_symbol: 'ETH',
    decimals: 18,
    logo_url:
      'https://static.debank.com/image/coin/logo_url/eth/6443cdccced33e204d90cb723c632917.png',
    protocol_id: '',
    price: 2753.63,
    price_24h_change: -0.004245363188289372,
    credit_score: 100000000.0,
    is_verified: true,
    is_scam: false,
    is_suspicious: false,
    is_core: true,
    is_wallet: true,
    time_at: 1622131200.0,
    amount: 0.001466057954290776,
  },
  pre_exec_version: 'v2',
  pre_exec: {
    success: true,
    error: null,
  },
  type_nft_send: {
    spender: '0x52DeC422AA4bAE0fB6F2b41db023dFfA9f4A1353',
    spender_protocol_name: null,
    spender_protocol_logo_url: '',
    token_symbol: 'Mission 200K: Arbitrum',
    token_amount: 1,
    is_infinity: false,
    is_nft: true,
    nft: {
      id: '7b31129c599fee1f53b8f439c0391072',
      contract_id: '0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
      inner_id: '440299',
      chain: 'arb',
      symbol: 'CUBE',
      name: 'Mission 200K: Arbitrum',
      description: '',
      content_type: 'image_url',
      content:
        'https://static.debank.com/image/arb_nft/local_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
      thumbnail_url:
        'https://static.debank.com/image/arb_nft/thumbnail_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
      total_supply: 1,
      attributes: [
        {
          trait_type: 'Type',
          value: 'Quest',
        },
        {
          trait_type: 'Title',
          value: 'Mission 200K: Arbitrum',
        },
        {
          trait_type: 'Community',
          value: 'Layer3',
        },
        {
          trait_type: 'Tag',
          value: 'DeFi',
        },
        {
          trait_type: 'Difficulty',
          value: 'BEGINNER',
        },
        {
          trait_type: 'Wallet Provider',
          value: 'Rabby Wallet',
        },
      ],
      detail_url:
        'https://opensea.io/assets/0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7/440299',
      collection_id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
      is_erc1155: false,
      is_erc721: true,
      pay_token: null,
      collection: {
        id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
        chain: 'arb',
        name: 'Layer3 CUBE',
        description: null,
        logo_url:
          'https://static.debank.com/image/arb_nft/thumbnail_url/24b44272ad6bf55323dca1add5f520f6/7b78944a6e25d3d2eded3eb6425802a1.png',
        is_verified: true,
        credit_score: 1643474.735188378,
        is_scam: false,
        is_suspicious: false,
        is_core: true,
        floor_price: 0.0,
      },
    },
  },
  trace_id: 'e01649fdaf9a43feb49cf907182e273d',
} as any;

export const parseTxData = {
  action: {
    type: 'send_nft',
    data: {
      to: '0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353',
      nft: {
        id: '7b31129c599fee1f53b8f439c0391072',
        contract_id: '0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
        inner_id: '440299',
        chain: 'arb',
        symbol: 'CUBE',
        name: 'Mission 200K: Arbitrum',
        description: '',
        content_type: 'image_url',
        content:
          'https://static.debank.com/image/arb_nft/local_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
        thumbnail_url:
          'https://static.debank.com/image/arb_nft/thumbnail_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png',
        total_supply: 1,
        attributes: [
          {
            trait_type: 'Type',
            value: 'Quest',
          },
          {
            trait_type: 'Title',
            value: 'Mission 200K: Arbitrum',
          },
          {
            trait_type: 'Community',
            value: 'Layer3',
          },
          {
            trait_type: 'Tag',
            value: 'DeFi',
          },
          {
            trait_type: 'Difficulty',
            value: 'BEGINNER',
          },
          {
            trait_type: 'Wallet Provider',
            value: 'Rabby Wallet',
          },
        ],
        detail_url:
          'https://opensea.io/assets/0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7/440299',
        collection_id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
        is_erc1155: false,
        is_erc721: true,
        pay_token: null,
        collection: {
          id: 'arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
          chain: 'arb',
          name: 'Layer3 CUBE',
          description: null,
          logo_url:
            'https://static.debank.com/image/arb_nft/thumbnail_url/24b44272ad6bf55323dca1add5f520f6/7b78944a6e25d3d2eded3eb6425802a1.png',
          is_verified: true,
          credit_score: 1643474.735188378,
          is_scam: false,
          is_suspicious: false,
          is_core: true,
          floor_price: 0.0,
        },
        amount: 1,
      },
    },
    is_asset_changed: true,
    is_involving_privacy: false,
  },
  contract_call: {
    func: 'safeTransferFrom',
    contract: {
      id: '0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
      protocol: {
        name: 'Layer3',
        logo_url:
          'https://static.debank.com/image/project/logo_url/layer3/17b9ec0fc10582671ac05d7429061b69.png',
      },
    },
  },
  log_id: 87962918,
} as any;

export const txData = {
  chainId: 42161,
  from: '0x5853ed4f26a3fcea565b3fbc698bb19cdf6deb85',
  to: '0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7',
  value: '0x0',
  data: '0x42842e0e0000000000000000000000005853ed4f26a3fcea565b3fbc698bb19cdf6deb8500000000000000000000000052dec422aa4bae0fb6f2b41db023dffa9f4a1353000000000000000000000000000000000000000000000000000000000006b7eb',
  gas: '',
  maxFeePerGas: '0xc65d40',
  maxPriorityFeePerGas: '0xc65d40',
  nonce: '0x6e3',
};
