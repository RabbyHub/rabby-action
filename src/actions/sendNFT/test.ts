import {
  apiProvider,
  ETH_CHAIN_ID,
  formatProvider,
  SENDER,
  walletProvider,
} from '../../../__mocks__';
import { ParsedTransactionActionData } from '../../types';
import { fetchDataSendNFT } from './fetchData';
import { formatSecurityEngineSendNFT } from './formatSecurityEngine';
import { parseTxData, preExecData, txData } from './mocks';
import { parseActionSendNFT } from './parseAction';
import {
  parseAction,
  fetchActionRequiredData,
  formatSecurityEngineContext,
} from '../..';

/**
 * Rabby Send NFT
 */
test.each([
  [parseActionSendNFT, fetchDataSendNFT, formatSecurityEngineSendNFT],
  [parseAction, fetchActionRequiredData, formatSecurityEngineContext],
])('SendNFT', async (_parseAction, _fetchData, _format) => {
  const actionData = _parseAction({
    type: 'transaction',
    data: parseTxData['action'],
    balanceChange: preExecData.balance_change,
    sender: txData.from,
    preExecVersion: preExecData.pre_exec_version,
    gasUsed: preExecData.gas.gas_used,
    tx: txData,
  }) as ParsedTransactionActionData;
  expect(actionData).toMatchSnapshot('parseActionSendNFT');

  const requireData = await _fetchData({
    type: 'transaction',
    actionData,
    contractCall: parseTxData.contract_call,
    chainId: ETH_CHAIN_ID,
    sender: SENDER,
    walletProvider,
    tx: txData,
    apiProvider,
  });

  expect(requireData).toMatchSnapshot('fetchDataSendNFT');

  const ctx = await _format({
    type: 'transaction',
    actionData,
    requireData,
    chainId: ETH_CHAIN_ID,
    isTestnet: false,
    provider: formatProvider,
  });
  expect(ctx).toMatchSnapshot('formatSecurityEngineSendNFT');
});
