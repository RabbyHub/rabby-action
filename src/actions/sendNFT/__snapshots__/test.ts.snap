// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendNFT: fetchDataSendNFT 1`] = `
{
  "cex": null,
  "contract": null,
  "eoa": {
    "bornAt": 1564655163,
    "id": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
  },
  "hasReceiverMnemonicInWallet": false,
  "hasReceiverPrivateKeyInWallet": true,
  "hasTransfer": true,
  "isTokenContract": false,
  "name": null,
  "onTransferWhitelist": false,
  "protocol": null,
  "receiverIsSpoofing": true,
  "usd_value": 4418.154405584272,
  "usedChains": [],
  "whitelistEnable": false,
}
`;

exports[`SendNFT: fetchDataSendNFT 2`] = `
{
  "cex": null,
  "contract": null,
  "eoa": {
    "bornAt": 1564655163,
    "id": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
  },
  "hasReceiverMnemonicInWallet": false,
  "hasReceiverPrivateKeyInWallet": true,
  "hasTransfer": true,
  "isTokenContract": false,
  "name": null,
  "onTransferWhitelist": false,
  "protocol": null,
  "receiverIsSpoofing": true,
  "usd_value": 4418.154405584272,
  "usedChains": [],
  "whitelistEnable": false,
}
`;

exports[`SendNFT: formatSecurityEngineSendNFT 1`] = `
{
  "sendNFT": {
    "cex": null,
    "chainId": "eth",
    "contract": null,
    "hasReceiverMnemonicInWallet": false,
    "hasReceiverPrivateKeyInWallet": true,
    "hasTransfer": true,
    "onTransferWhitelist": false,
    "receiverIsSpoofing": true,
    "to": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
    "usedChainList": [],
  },
}
`;

exports[`SendNFT: formatSecurityEngineSendNFT 2`] = `
{
  "sendNFT": {
    "cex": null,
    "chainId": "eth",
    "contract": null,
    "hasReceiverMnemonicInWallet": false,
    "hasReceiverPrivateKeyInWallet": true,
    "hasTransfer": true,
    "onTransferWhitelist": false,
    "receiverIsSpoofing": true,
    "to": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
    "usedChainList": [],
  },
}
`;

exports[`SendNFT: parseActionSendNFT 1`] = `
{
  "sendNFT": {
    "nft": {
      "amount": 1,
      "attributes": [
        {
          "trait_type": "Type",
          "value": "Quest",
        },
        {
          "trait_type": "Title",
          "value": "Mission 200K: Arbitrum",
        },
        {
          "trait_type": "Community",
          "value": "Layer3",
        },
        {
          "trait_type": "Tag",
          "value": "DeFi",
        },
        {
          "trait_type": "Difficulty",
          "value": "BEGINNER",
        },
        {
          "trait_type": "Wallet Provider",
          "value": "Rabby Wallet",
        },
      ],
      "chain": "arb",
      "collection": {
        "chain": "arb",
        "credit_score": 1643474.735188378,
        "description": null,
        "floor_price": 0,
        "id": "arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
        "is_core": true,
        "is_scam": false,
        "is_suspicious": false,
        "is_verified": true,
        "logo_url": "https://static.debank.com/image/arb_nft/thumbnail_url/24b44272ad6bf55323dca1add5f520f6/7b78944a6e25d3d2eded3eb6425802a1.png",
        "name": "Layer3 CUBE",
      },
      "collection_id": "arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
      "content": "https://static.debank.com/image/arb_nft/local_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png",
      "content_type": "image_url",
      "contract_id": "0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
      "description": "",
      "detail_url": "https://opensea.io/assets/0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7/440299",
      "id": "7b31129c599fee1f53b8f439c0391072",
      "inner_id": "440299",
      "is_erc1155": false,
      "is_erc721": true,
      "name": "Mission 200K: Arbitrum",
      "pay_token": null,
      "symbol": "CUBE",
      "thumbnail_url": "https://static.debank.com/image/arb_nft/thumbnail_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png",
      "total_supply": 1,
    },
    "to": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
  },
}
`;

exports[`SendNFT: parseActionSendNFT 2`] = `
{
  "sendNFT": {
    "nft": {
      "amount": 1,
      "attributes": [
        {
          "trait_type": "Type",
          "value": "Quest",
        },
        {
          "trait_type": "Title",
          "value": "Mission 200K: Arbitrum",
        },
        {
          "trait_type": "Community",
          "value": "Layer3",
        },
        {
          "trait_type": "Tag",
          "value": "DeFi",
        },
        {
          "trait_type": "Difficulty",
          "value": "BEGINNER",
        },
        {
          "trait_type": "Wallet Provider",
          "value": "Rabby Wallet",
        },
      ],
      "chain": "arb",
      "collection": {
        "chain": "arb",
        "credit_score": 1643474.735188378,
        "description": null,
        "floor_price": 0,
        "id": "arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
        "is_core": true,
        "is_scam": false,
        "is_suspicious": false,
        "is_verified": true,
        "logo_url": "https://static.debank.com/image/arb_nft/thumbnail_url/24b44272ad6bf55323dca1add5f520f6/7b78944a6e25d3d2eded3eb6425802a1.png",
        "name": "Layer3 CUBE",
      },
      "collection_id": "arb:0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
      "content": "https://static.debank.com/image/arb_nft/local_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png",
      "content_type": "image_url",
      "contract_id": "0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7",
      "description": "",
      "detail_url": "https://opensea.io/assets/0x1195cf65f83b3a5768f3c496d3a05ad6412c64b7/440299",
      "id": "7b31129c599fee1f53b8f439c0391072",
      "inner_id": "440299",
      "is_erc1155": false,
      "is_erc721": true,
      "name": "Mission 200K: Arbitrum",
      "pay_token": null,
      "symbol": "CUBE",
      "thumbnail_url": "https://static.debank.com/image/arb_nft/thumbnail_url/7b31129c599fee1f53b8f439c0391072/f2830449700b7dc22ac04f0caf54b1f8.png",
      "total_supply": 1,
    },
    "to": "0x52dec422aa4bae0fb6f2b41db023dffa9f4a1353",
  },
}
`;
